tensorflow/go/operation.go

```diff
@@ -123,6 +148,67 @@ func (p Output) c() C.TF_Output {
 
 func (p Output) canBeAnInput() {}
 
+// NumConsumers returns the number of inputs that consume this output.
+func (p Output) NumConsumers() int {
+	return int(C.TF_OperationOutputNumConsumers(p.c()))
+}
+
+// Consumers returns the inputs that consume this output.
+func (p Output) Consumers() []OpInput {
+	max := p.NumConsumers()
+	consumers := make([]C.TF_Input, max)
+	n := C.TF_OperationOutputConsumers(p.c(), (*C.TF_Input)(unsafe.Pointer(&consumers[0])), C.int(max))
+	consumers = consumers[:int(n)]
+
+	var inputs []OpInput
+	for _, consumer := range consumers {
+		inputs = append(inputs, OpInput{
+			Index: int(consumer.index),
+			Op: &Operation{
+				c: consumer.oper,
+				g: p.Op.g,
+			},
+		})
+	}
+
+	return inputs
+}
+
+// OpInput is one of the inputs to an operation.
+type OpInput struct {
+	// Op is the Operation that consumes this Input.
+	Op *Operation
+
+	// Index specifies the index of the input within the Operation.
+	Index int
+}
+
+func (p OpInput) c() C.TF_Input {
+	if p.Op == nil {
+		// Attempt to provide a more useful panic message than "nil
+		// pointer dereference".
+		panic("nil-Operation. If the Output was created with a Scope object, see Scope.Err() for details.")
+	}
+	return C.TF_Input{oper: p.Op.c, index: C.int(p.Index)}
+}
+
+// DataType returns the type of elements in the tensor produced by p.
```

Comment:
This comment doesn't seem right, `p` doesn't "produce" a tensor.
Also, it seems that the input datatype here must match the datatype of the `Output` object from which this `OpInput` was obtained, right?

If so, would it make sense to not expose this and instead have people rely on the single way to get the appropriate `DataType` - from `Output.DataType()`? Keeping with the theme of keeping the API surface small.

Again though, I'm open to keeping this if there is a strong use-case.

Reply:
I think it's helpful even if it's just a helper method. If you're walking the graph upwards it might simplify the code at each node. Is there any way to create a node with an unlinked input via Python/C?

Reply:
So the simplification now is over `p.Producer().DataType()`?
I'm fine either way (keeping or removing it)

Reply:
Leaving it in seems easiest